File: src/wy_qcos/transpiler/cmss/optimizer/inverse_cancellation.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
# ----------------------------------------------------------------------
# Copyright© 2024-2026 China Mobile (SuZhou) Software Technology Co.,Ltd.
#
# qcos is licensed under Mulan PSL v2.
# You can use this software according to the terms and conditions
# of the Mulan PSL v2.
# You may obtain a copy of Mulan PSL v2 at:
#         http://license.coscl.org.cn/MulanPSL2
# THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS,
#     WITHOUT WARRANTIES OF ANY KIND,
# EITHER EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT,
# MERCHANTABILITY OR FIT FOR A PARTICULAR PURPOSE.
# See the Mulan PSL v2 for more details.
# ----------------------------------------------------------------------

import numpy as np

from wy_qcos.transpiler.cmss.common.gate_operation import GateOperation
from wy_qcos.transpiler.cmss.circuit.dag_circuit import DAGCircuit


class InverseCancellation:
    """Cancel inverse gates.

    Cancel specific Gates which are inverses of each other when they occur
    back-to-back.
    """

    def __init__(
        self,
        gates_to_cancel: list[
            GateOperation | tuple[GateOperation, GateOperation]
        ],
    ):
        for gates in gates_to_cancel:
            if isinstance(gates, GateOperation):
                if not self._is_inverse(gates):
                    raise ValueError(f"Gate {gates.name} is not self-inverse")
            elif isinstance(gates, tuple):
                if len(gates) != 2:
                    raise ValueError(
                        f"Too many or too few inputs: {gates}. \
                          Only two are allowed."
                    )
                if not self._is_inverse(gates[0], gates[1]):
                    raise ValueError(
                        f"Gate {gates[0].name} and {gates[1].name} \
                          are not inverse."
                    )
            else:
                raise ValueError(
                    f"InverseCancellation pass does not take input \
                      type {type(gates)}. Input must be a Gate."
                )

        self.self_inverse_gates = []
        self.inverse_gate_pairs = []
        self.self_inverse_gate_names = set()
        self.inverse_gate_pairs_names = set()

        for gates in gates_to_cancel:
            if isinstance(gates, GateOperation):
                self.self_inverse_gates.append(gates)
                self.self_inverse_gate_names.add(gates.name)
            else:
                self.inverse_gate_pairs.append(gates)
                self.inverse_gate_pairs_names.update(x.name for x in gates)

    def run(self, dag: DAGCircuit):
        """Run the InverseCancellation pass on `dag`.

        Args:
            dag: the directed acyclic graph to run on.

        Returns:
            DAGCircuit: Transformed DAG.
        """
        if self.self_inverse_gates:
            dag = self._run_on_self_inverse(dag)
        if self.inverse_gate_pairs:
            dag = self._run_on_inverse_pairs(dag)
        return dag

    def _is_inverse(
        self, gate1: GateOperation, gate2: GateOperation | None = None
    ) -> bool:
        """Check if two gates are inverses of each other.

        If `gate2` is None, this function checks whether `gate1` is
        self-inverse. Otherwise, this function checks whether `gate1` and
        `gate2` are inverses. Global phase is not considered. This is safe
        because the input gates are assumed to be basic gates.

        Args:
            gate1 (GateOperation): first gate.
            gate2 (GateOperation, optional): second gate. Defaults to None.

        Returns:
            bool: inverse or not.
        """
        op1 = gate1.to_matrix()
        if gate2 is None:
            op2 = op1
        else:
            op2 = gate2.to_matrix()

        if op1.shape != op2.shape:
            return False

        product = np.matmul(op1, op2)
        dim = op1.shape[0]
        identity = np.eye(dim, dtype=complex)

        return np.allclose(product, identity)

    def _run_on_self_inverse(self, dag: DAGCircuit):
        """Run self-inverse gates on `dag`.

        Args:
            dag: the directed acyclic graph to run on.

        Returns:
            DAGCircuit: Transformed DAG.
        """
        op_counts = dag.count_ops()
        if not self.self_inverse_gate_names.intersection(op_counts):
            return dag

        for gate in self.self_inverse_gates:
            gate_name = gate.name
            gate_count = op_counts.get(gate_name, 0)
            if gate_count <= 1:
                continue
            gate_runs = dag.collect_runs([gate_name])
            for gate_cancel_run in gate_runs:
                partitions = []
                chunk = []
                max_index = len(gate_cancel_run) - 1
                for i in range(len(gate_cancel_run)):
                    if gate_cancel_run[i].op.name == gate_name:
                        chunk.append(gate_cancel_run[i])
                    else:
                        if chunk:
                            partitions.append(chunk)
                            chunk = []
                        continue
                    # check qargs, cx[0, 1] and cx[1, 0] cannot cancel
                    if (
                        i == max_index
                        or gate_cancel_run[i].qargs
                        != gate_cancel_run[i + 1].qargs
                    ):
                        partitions.append(chunk)
                        chunk = []
                # Remove an even number of gates from each chunk
                for chunk in partitions:
                    if len(chunk) % 2 == 0:
                        dag.remove_op_node(chunk[0])
                    for node in chunk[1:]:
                        dag.remove_op_node(node)
        return dag

    def _run_on_inverse_pairs(self, dag: DAGCircuit):
        """Run inverse gate pairs on `dag`.

        Args:
            dag: the directed acyclic graph to run on.

        Returns:
            DAGCircuit: Transformed DAG.
        """
        op_counts = dag.count_ops()
        if not self.inverse_gate_pairs_names.intersection(op_counts):
            return dag

        for pair in self.inverse_gate_pairs:
            gate_0_name = pair[0].name
            gate_1_name = pair[1].name
            if gate_0_name not in op_counts or gate_1_name not in op_counts:
                continue
            gate_cancel_runs = dag.collect_runs([gate_0_name, gate_1_name])
            for dag_nodes in gate_cancel_runs:
                i = 0
                while i < len(dag_nodes) - 1:
                    if (
                        dag_nodes[i].qargs == dag_nodes[i + 1].qargs
                        and dag_nodes[i].op.name == pair[0].name
                        and dag_nodes[i + 1].op.name == pair[1].name
                    ):
                        dag.remove_op_node(dag_nodes[i])
                        dag.remove_op_node(dag_nodes[i + 1])
                        i = i + 2
                    elif (
                        dag_nodes[i].qargs == dag_nodes[i + 1].qargs
                        and dag_nodes[i].op.name == pair[1].name
                        and dag_nodes[i + 1].op.name == pair[0].name
                    ):
                        dag.remove_op_node(dag_nodes[i])
                        dag.remove_op_node(dag_nodes[i + 1])
                        i = i + 2
                    else:
                        i = i + 1
        return dag

File: src_cpp/transpiler_cpp/src/optimizer/inverse_cancellation.cpp
/*
 * ----------------------------------------------------------------------
 * Copyright© 2024-2026 China Mobile (SuZhou) Software Technology Co.,Ltd.
 *
 * qcos is licensed under Mulan PSL v2.
 * You can use this software according to the terms and conditions
 * of the Mulan PSL v2.
 * You may obtain a copy of Mulan PSL v2 at:
 *          http://license.coscl.org.cn/MulanPSL2
 * THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS,
 *      WITHOUT WARRANTIES OF ANY KIND,
 * EITHER EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT,
 * MERCHANTABILITY OR FIT FOR A PARTICULAR PURPOSE.
 * See the Mulan PSL v2 for more details.
 * ----------------------------------------------------------------------
 */

#include "optimizer/inverse_cancellation.h"

#include <iostream>
#include <stdexcept>
#include <unordered_map>

#include "circuit/dag_node.h"

namespace qcos {

InverseCancellation::InverseGateRule::InverseGateRule(GateOperation gate)
    : gate_0(std::move(gate)) {}

InverseCancellation::InverseGateRule::InverseGateRule(GateOperation gate_0_in,
                                                      GateOperation gate_1_in)
    : gate_0(std::move(gate_0_in)), gate_1(std::move(gate_1_in)) {}

InverseCancellation::InverseCancellation(
    const std::vector<InverseGateRule>& gates_to_cancel, bool verbose)
    : verbose_(verbose) {
  for (const auto& gates : gates_to_cancel) {
    // 判断是否为自反门
    if (!gates.is_pair()) {
      // 判断 gate_0 是否为自反门
      if (!is_inverse(gates.gate_0)) {
        throw std::invalid_argument("Gate " + gates.gate_0.name +
                                    " is not self-inverse");
      }
      self_inverse_gate_names_.insert(gates.gate_0.name);
      continue;
    }

    // 判断 gate_0 和 gate_1 是否互逆
    if (!is_inverse(gates.gate_0, gates.gate_1)) {
      throw std::invalid_argument("Gate " + gates.gate_0.name + " and " +
                                  gates.gate_1->name + " are not inverse.");
    }
    inverse_gate_pairs_.push_back(gates);
    inverse_gate_pairs_names_.insert(gates.gate_0.name);
    inverse_gate_pairs_names_.insert(gates.gate_1->name);
  }
}

int InverseCancellation::run(
    DAGCircuit& dag, const std::optional<std::set<std::string>>& basis_gates) {
  int reduced = 0;
  // 入口算一次拓扑序，供所有规则复用，避免每次 collect_runs 重复排序
  const auto topo_order = dag.get_multi_graph().topo_order();
  if (!self_inverse_gate_names_.empty()) {
    // 处理自反门的成对消除逻辑
    reduced += run_on_self_inverse(dag, topo_order);
  }
  if (!inverse_gate_pairs_.empty()) {
    // 处理互逆门的成对消除逻辑
    reduced += run_on_inverse_pairs(dag, topo_order);
  }
  if (verbose_) {
    std::clog << name() << ": " << reduced << " gates reduced\n";
  }
  return reduced;
}

bool InverseCancellation::is_inverse(
    const GateOperation& gate_0,
    const std::optional<GateOperation>& gate_1) const {
  if (!gate_1) {
    return gate_0.hermitian;
  }
  if (gate_0.operation_type != gate_1->operation_type) {
    return false;
  }
  if (gate_0.name == gate_1->name) {
    return gate_0.hermitian;
  }

  // 目前仅支持 s, sdg, t, tdg 这两对显式互逆门
  // 其他门可通过设置 hermitian=true 来实现自反门对消。
  static const std::unordered_map<std::string, std::string> kInversePairs = {
      {"s", "sdg"}, {"sdg", "s"}, {"t", "tdg"}, {"tdg", "t"}};

  auto iterator = kInversePairs.find(gate_0.name);
  return iterator != kInversePairs.end() && iterator->second == gate_1->name;
}

int InverseCancellation::run_on_self_inverse(
    DAGCircuit& dag, const std::vector<int>& topo_order) const {
  const auto op_counts = dag.count_ops();
  int reduced = 0;

  for (const auto& gate_name : self_inverse_gate_names_) {
    auto count_iterator = op_counts.find(gate_name);
    // 如果电路中没有这个门，或者这个门只出现了一次（无法成对消去），则跳过。
    if (count_iterator == op_counts.end() || count_iterator->second <= 1) {
      continue;
    }

    // 收集所有同名门的连续串
    auto gate_runs = dag.collect_runs({gate_name}, &topo_order);
    for (const auto& gate_cancel_run : gate_runs) {
      // partition 用于将连续串切分成若干段，每段内门的 qargs 都相同
      std::vector<std::vector<DAGNode*>> partitions;
      // chunk 用于临时存储当前段的节点，遇到 qargs 变化或串尾时就切分出一段。
      std::vector<DAGNode*> chunk;
      const size_t max_index =
          gate_cancel_run.empty() ? 0u : gate_cancel_run.size() - 1;

      for (size_t index = 0; index < gate_cancel_run.size(); ++index) {
        auto* current = dynamic_cast<DAGOpNode*>(gate_cancel_run[index]);
        if (!current || current->name() != gate_name) {
          if (!chunk.empty()) {
            partitions.push_back(chunk);
            chunk.clear();
          }
          continue;
        }

        chunk.push_back(gate_cancel_run[index]);
        auto* next =
            index == max_index
                ? nullptr
                : dynamic_cast<DAGOpNode*>(gate_cancel_run[index + 1]);
        // 当 qargs
        // 变化时切分出一段，保证每段内的门都作用在同一组量子位上，才可成对消去。
        const bool qargs_changed =
            index == max_index || !next || current->qargs != next->qargs;
        if (qargs_changed) {
          partitions.push_back(chunk);
          chunk.clear();
        }
      }

      for (const auto& partition : partitions) {
        // 如果分区内的门数量为奇数，则保留第一个门，其余门成对消去。
        const size_t keep_prefix = partition.size() % 2 == 0 ? 0u : 1u;
        for (size_t index = keep_prefix; index < partition.size(); ++index) {
          dag.remove_op_node(dynamic_cast<DAGOpNode*>(partition[index]));
          ++reduced;
        }
      }
    }
  }

  return reduced;
}

int InverseCancellation::run_on_inverse_pairs(
    DAGCircuit& dag, const std::vector<int>& topo_order) const {
  const auto op_counts = dag.count_ops();
  int reduced = 0;

  for (const auto& pair : inverse_gate_pairs_) {
    const auto& gate_0_name = pair.gate_0.name;
    const auto& gate_1_name = pair.gate_1->name;
    if (op_counts.find(gate_0_name) == op_counts.end() ||
        op_counts.find(gate_1_name) == op_counts.end()) {
      continue;
    }

    auto gate_cancel_runs =
        dag.collect_runs({gate_0_name, gate_1_name}, &topo_order);
    for (const auto& dag_nodes : gate_cancel_runs) {
      size_t index = 0;
      while (index + 1 < dag_nodes.size()) {
        auto* first = dynamic_cast<DAGOpNode*>(dag_nodes[index]);
        auto* second = dynamic_cast<DAGOpNode*>(dag_nodes[index + 1]);
        if (!first || !second) {
          ++index;
          continue;
        }

        // 只有当两个门作用在同一组量子位上时才可成对消去。
        const bool same_qargs = first->qargs == second->qargs;
        const bool ordered_match = same_qargs &&
                                   first->name() == gate_0_name &&
                                   second->name() == gate_1_name;
        const bool reverse_match = same_qargs &&
                                   first->name() == gate_1_name &&
                                   second->name() == gate_0_name;
        if (ordered_match || reverse_match) {
          dag.remove_op_node(first);
          dag.remove_op_node(second);
          reduced += 2;
          index += 2;
          continue;
        }
        ++index;
      }
    }
  }

  return reduced;
}

}  // namespace qcos
